lib/Dialect/ESI/ESIServices.cpp

```diff
@@ -68,33 +69,42 @@ static LogicalResult instantiateCosimEndpointOps(ServiceImplementReqOp implReq,
     return StringAttr::get(ctxt, os.str());
   };
 
-  llvm::DenseMap<ServiceReqOpInterface, unsigned> toClientResultNum;
-  for (auto req : implReq.getOps<ServiceReqOpInterface>())
-    if (req.getToClient())
-      toClientResultNum[req] = toClientResultNum.size();
+  llvm::DenseMap<ServiceImplementConnReqOp, unsigned> toClientResultNum;
+  for (auto req : implReq.getOps<ServiceImplementConnReqOp>())
+    toClientResultNum[req] = toClientResultNum.size();
 
-  // Iterate through them, building a cosim endpoint for each one.
-  for (auto req : implReq.getOps<ServiceReqOpInterface>()) {
+  // Iterate through the requests, building a cosim endpoint for each channel in
+  // the bundle.
+  // TODO: The cosim op should probably be able to take a bundle type and get
+  // lowered to the SV primitive later on. The SV primitive will also need some
+  // work to suit this new world order, so let's put this off.
+  for (auto req : implReq.getOps<ServiceImplementConnReqOp>()) {
     Location loc = req->getLoc();
+    ChannelBundleType bundleType = req.getToClient().getType();
     ArrayAttr clientNamePathAttr = req.getClientNamePath();
 
-    Value toServerValue = req.getToServer();
-    if (!toServerValue)
-      toServerValue =
-          b.create<NullSourceOp>(loc, ChannelType::get(ctxt, b.getI1Type()));
+    SmallVector<Value, 8> toServerValues;
+    for (BundledChannel ch : bundleType.getChannels()) {
+      if (ch.direction == ChannelDirection::to) {
+        auto nullSource = b.create<NullSourceOp>(loc, i1ch);
+        auto cosim = b.create<CosimEndpointOp>(
+            loc, ch.type, clk, rst, nullSource.getOut(),
+            toStringAttr(clientNamePathAttr));
+        toServerValues.push_back(cosim.getRecv());
+      }
+    }
 
-    Type toClientType = req.getToClientType();
-    if (!toClientType)
-      toClientType = ChannelType::get(ctxt, b.getI1Type());
+    auto pack =
+        b.create<PackBundleOp>(implReq.getLoc(), bundleType, toServerValues);
+    implReq.getResult(toClientResultNum[req])
+        .replaceAllUsesWith(pack.getBundle());
 
-    auto cosim =
-        b.create<CosimEndpointOp>(loc, toClientType, clk, rst, toServerValue,
+    size_t chanIdx = 0;
+    for (BundledChannel ch : bundleType.getChannels())
+      if (ch.direction == ChannelDirection::from)
+        b.create<CosimEndpointOp>(loc, i1ch, clk, rst,
+                                  pack.getFromChannels()[chanIdx++],
                                   toStringAttr(clientNamePathAttr));
```

Comment:
really not a big fan of unbracketed for loops when the inner nested statement breaks over multiple lines. [technically not against the style guidelines](https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements) although one can argue whether this counts as "more than two levels of nesting".